---
title: "05_network_reconstruction_with_CARNIVAL"
author: "Rosa Hernansaiz-Ballesteros"
date: "18/05/2020"
output: 
  github_document: default
  
---

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### License Info

This program is free software: you can redistribute it and/or modify it under 
the terms of the GNU General Public License as published by the Free Software 
Foundation, either version 3 of the License, or (at your option) any later 
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY 
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR 
A PARTICULAR PURPOSE. See the GNU General Public License for more details.

Please check http://www.gnu.org/licenses/.

## Introduction

**CARNIVAL** (CAusal Reasoning for Network identification using Integer VALue
programming) is a method for the identification of upstream regulatory
signalling pathways from downstream gene expression (GEX) [(Liu, Trairatphisan,
Gjerga et al. 2019)](https://doi.org/10.1038/s41540-019-0118-z). The aim of
**CARNIVAL** is to identify a subset of interactions from a prior knowledge
network, a network that represents potential regulated pathways linking known or
potential targets of perturbation, to active transcription factors derived
from GEX data.

In the fifth part of our transcriptomics tutorial series, we demonstrate how to
use **CARNIVAL** based on the transcription factor (TF) activity derived from
transcriptomics data using **DoRothEA** [(Garcia-Alonso et al.
2019)](https://doi.org/10.1101/gr.240663.118), and the prior knowledge network
obtained from [**Omnipath**](http://omnipathdb.org/), a literature curated
mammalian signaling pathway resource [(TÃ¼rei et al
2016)](https://www.nature.com/articles/nmeth.4077).

In order to help CARNIVAL to find a solution faster,
we can also use **PROGENy** [(Schubert et al. 2018)](https://www.nature.com/articles/s41467-017-02391-6) scores
to infer the score of the representative genes for each of the calculated pathways.

[**CARNIVAL**](https://www.bioconductor.org/packages/release/bioc/html/CARNIVAL.html),
[**DoRothEA**](http://bioconductor.org/packages/release/data/experiment/html/dorothea.html),
[**PROGENy**](http://bioconductor.org/packages/release/bioc/html/progeny.html) and
[**OmnipathR**](https://www.bioconductor.org/packages/release/bioc/html/OmnipathR.html)
are available as bioconductor packages.
Visit our website for additional information about each of the tools. https://saezlab.org

## Getting Started

We first load the required libraries and the support functions.

```{r libraries, message=FALSE}
library(progeny)
library(dorothea)
library(CARNIVAL)
library(OmnipathR)
library(readr)
library(tibble)
library(tidyr)
library(dplyr)
library(visNetwork)

library(ggplot2)
library(pheatmap)


## For the volcano plot (related to support functions)
library(ggrepel)

## We also load the support functions
source("assignPROGENyScores.r")
source("generateTFList.r")
source("carnival_visNetwork.r")
```

In addition, we read the results from the previous scripts: 

+ Transcription factor activities (04_TranscriptionFactor_activity_with_Dorothea.Rmd)
+ Pathways activity scores (03_Pathway_activity_with_Progeny.Rmd). 
 

```{r loadInput, message=FALSE}
## We read the normalised counts and the experimental design 
tf_activities <- read_csv("../results/TFActivity_CARNIVALinput.csv")
PathwayActivity <- read_csv("../results/PathwayActivity_CARNIVALinput.csv")
```

## Getting the scaffold network from Omnipath

Before running **CARNIVAL**, we need to create or upload a scaffold network.
This will be _"the map"_ that the ILP algorithm will follow to find the causal network.
We use **Omnipath** to obtain the signed and directed interactions from all the available resources.
CARNIVAL requires this information in a _sif_ table (node1, interaction, node2) format,
therefore we use the _consensus_ columns of direction (consensus_direction) and
sign (consensus_stimulation and consensus_inhibition) to extract it.

The query returns 0/1 as logic status of being a stimulation or an inhibition reaction. 
Thus, this output is reformulated as 1/-1 to indicate stimulation or inhibition, respectively.
We can keep either the interactions that are consistent, or both alternatives (e.g. A 1 B; A -1 B).
In this example, we keep the consistent ones.

```{r omnipathSIF, message=FALSE}

omniR <- import_Omnipath_Interactions()
  
# signed and directed
omnipath_sd <- omniR %>% dplyr::filter(consensus_direction == 1 &
                                (consensus_stimulation == 1 | 
                                 consensus_inhibition == 1
                                 ))
  
# changing 0/1 criteria in consensus_stimulation/inhibition to -1/1
omnipath_sd$consensus_stimulation[which( omnipath_sd$consensus_stimulation == 0)] = -1
omnipath_sd$consensus_inhibition[which( omnipath_sd$consensus_inhibition == 1)] = -1
omnipath_sd$consensus_inhibition[which( omnipath_sd$consensus_inhibition == 0)] = 1

# check consistency on consensus sign and select only those in a SIF format
sif <- omnipath_sd[,c('source_genesymbol', 'consensus_stimulation', 'consensus_inhibition', 'target_genesymbol')] %>%
      dplyr::filter(consensus_stimulation==consensus_inhibition) %>%
      unique.data.frame()

sif$consensus_stimulation <- NULL
colnames(sif) <- c('source', 'interaction', 'target')

# remove complexes
sif$source <- gsub(":", "_", sif$source)
sif$target <- gsub(":", "_", sif$target)

#save SIF
write_tsv(sif, "../results/omnipath_carnival.tsv")

```

## Transcription Factor and pathway activities for CARNIVAL

We use the supplementary functions _generateTFList.r_ and _assignPROGENyScores.r_
to shift the formats of tf_activities and PathwayActivity to the one required by CARNIVAL.

```{r shiftFormats, message=FALSE}

# dorothea for CARNIVAL
tf_activities_carnival <- data.frame(tf_activities, stringsAsFactors = F)
rownames(tf_activities_carnival) <- tf_activities$TF
tf_activities_carnival$TF <- NULL
tfList = generateTFList(tf_activities_carnival, top=50, access_idx = 1)

# progeny for CARNIVAL
load(file = system.file("progenyMembers.RData",package="CARNIVAL"))

PathwayActivity_carnival <- data.frame(PathwayActivity, stringsAsFactors = F)
rownames(PathwayActivity_carnival) <- PathwayActivity_carnival$Pathway
PathwayActivity_carnival$Pathway <- NULL
progenylist = assignPROGENyScores(progeny = t(PathwayActivity_carnival), 
                                            progenyMembers = progenyMembers, 
                                            id = "gene", 
                                            access_idx = 1)
```

## Running CARNIVAL

CARNIVAL has been developed to find the causal link between the activities of the transcription factors (TFs) and the 'perturbed' nodes.
In current version, v1.0.0, we have 3 main inputs that we have to provide:

+ _measObj_: The TFs' activities (like the ones we have obtained from DoRothEA)
+ _inputObj_: The 'perturbed' nodes we want that CARNIVAL connects with the activity of TFs. 
There are 3 ways of using it:
(1) Give the name and sign of the selected nodes; 
(2) Give the name only, so the algorithm will select the sign that best fit the models,
(3) Give _NULL_ as value will create a "Perturbation" node that will try both signs for all 'initial' nodes of the given network ( _netObj_ ).
+ _netObj_: The network that will serve as map to connect the TFs' activities ( _measObj_ ) and the perturbed nodes ( _inputObj_ )

Although it is not required, a fourth object called _weightObj_ can be also given. 
This object gives values ranged from -1 to 1 for a set of nodes of the network.
The aim of _weightObj_ is helping the solver to find optimal solutions faster.

In the present example, we use assign as perturbation nodes all the "initial" nodes (option 2),
and as _weightObj_ the PROGENy scores assigned to the most representative genes of the calculated pathways,

Please, check the [CARNIVAL](https://saezlab.github.io/CARNIVAL/) page to get some more insight.
For more specific technical details, as how to install different ILP solvers that CARNIVAL supports,
pleasre check [CARNIVAL vignette](https://bioconductor.org/packages/release/bioc/vignettes/CARNIVAL/inst/doc/CARNIVAL.html).

```{r carnival, message=FALSE}

# get initial nodes
iniMTX = base::setdiff(sif$source, sif$target)
iniciators = base::data.frame(base::matrix(data = NaN, nrow = 1, ncol = length(iniMTX)), stringsAsFactors = F)
colnames(iniciators) = iniMTX

# run carnival
carnival_result = runCARNIVAL( inputObj= iniciators,
                               measObj = tfList$t, 
                               netObj = sif, 
                               weightObj = progenylist$score, 
                               solverPath = "/Applications/CPLEX_Studio129/cplex/bin/x86-64_osx/cplex", 
                               solver = "cplex",
                               timelimit=7200,
                               mipGAP=0,
                               poolrelGAP=0 )
```

CARNIVAL gives a list of 4 elements:

+ weightedSIF: summary of all interactions found in all models
+ nodesAttributes: summary of all nodes and how many times are they found in the different models
+ sifAll: networks of all the models
+ attributesAll: node attributes of all models

We can now visualise the network...

```{r visNetwork1, message=FALSE, dpi=300}

#transoform to data.frame
carnival_result$weightedSIF <- data.frame(carnival_result$weightedSIF, stringsAsFactors = F)
carnival_result$weightedSIF$Sign <- as.numeric(carnival_result$weightedSIF$Sign)
carnival_result$weightedSIF$Weight <- as.numeric(carnival_result$weightedSIF$Weight)

carnival_result$nodesAttributes <- data.frame(carnival_result$nodesAttributes, stringsAsFactors = F)
carnival_result$nodesAttributes$ZeroAct <- as.numeric(carnival_result$nodesAttributes$ZeroAct)
carnival_result$nodesAttributes$UpAct <- as.numeric(carnival_result$nodesAttributes$UpAct)
carnival_result$nodesAttributes$DownAct <- as.numeric(carnival_result$nodesAttributes$DownAct)
carnival_result$nodesAttributes$AvgAct <- as.numeric(carnival_result$nodesAttributes$AvgAct)

saveRDS(carnival_result,"../results/carnival_result.rds")

# visualization
visNet = carnival_visNet(evis = carnival_result$weightedSIF,
                         nvis = carnival_result$nodesAttributes)
#visNet
visSave(visNet, file = paste0('carnival_visualization_visNetwork.html'), selfcontained = TRUE)

```


## References

> Liu A., Trairatphisan P., Gjerga E. et al. "From expression footprints to causal pathways: contextualizing large signaling networks with CARNIVAL". _npj Systems Biology and Applications_. 2019. DOI: [10.1038/s41540-019-0118-z](https://www.nature.com/articles/s41540-019-0118-z)

> Garcia-Alonso L, Holland CH, Ibrahim MM, Turei D, Saez-Rodriguez J. âBenchmark and integration of resources for the estimation of human transcription factor activities.â _Genome Research_. 2019. DOI: 
[10.1101/gr.240663.118](https://genome.cshlp.org/content/29/8/1363).

> TÃ¼rei D, KorcsmÃ¡ros T, & Saez-Rodriguez J, "OmniPath: guidelines and gateway for literature-curated signaling pathway resources". _Nat Methods_ 2016. DOI: [10.1038/nmeth.4077](https://doi.org/10.1038/nmeth.4077).

> Schubert M, Klinger B, KlÃ¼nemann M, Sieber A, Uhlitz F, Sauer S, Garnett MJ, BlÃ¼thgen N, Saez-Rodriguez J. âPerturbation-response genes reveal signaling footprints in cancer gene expression.â _Nature Communications_. 2018. DOI: [10.1038/s41467-017-02391-6](https://doi.org/10.1038/s41467-017-02391-6)


## Session Info Details

```{r sessionInfo, echo=FALSE, eval=TRUE}
sessionInfo()
```
